### PR TITLE
[Impeller] Will it blend?

### PR DESCRIPTION
**Pull request by bdero**
That is the question... and the answer is yes.

Adds all the remaining "separable" blend modes. I plan to land the 4 remaining ("non-separable") blends (Hue, Saturation, Color, and Luminosity) in a separate patch.

https://user-images.githubusercontent.com/919017/171968607-d69585f4-30f7-4abc-916b-03011ee83e91.mov

### CODE REVIEW

File: impeller/entity/shaders/blending/advanced_blend_hardlight.frag [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "advanced_blend_utils.glsl"
+
+vec3 Blend(vec3 dst, vec3 src) {
+  return BlendHardLight(dst, src);
+}
+
+#include "advanced_blend.glsl"
```
Comment (chinmaygarde) [inline impeller/entity/shaders/blending/advanced_blend_hardlight.frag:11]: Nit: Newline at EOF

File: impeller/entity/shaders/blending/advanced_blend_softlight.frag [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "advanced_blend_utils.glsl"
+
+vec3 Blend(vec3 dst, vec3 src) {
+  vec3 D = MixComponents(((16 * dst - 12) * dst + 4) * dst,  //
```
Comment (chinmaygarde) [inline impeller/entity/shaders/blending/advanced_blend_softlight.frag:8]: I had to look up which variant you'd used on Wikipedia. I think it'd be good to document this is W3C variant here. Adding documentation for all blend modes either inline here or in a new doc in the `docs/` folder would be great (in a separate patch is fine).

**Review by chinmaygarde — APPROVED** [review]
This is great. Left a few comments inline but they are mostly nits.

File: impeller/entity/shaders/blending/advanced_blend_hardlight.frag (hunk 1 above)
Comment (bdero) [reply impeller/entity/shaders/blending/advanced_blend_hardlight.frag:11]: Done.

File: impeller/entity/shaders/blending/advanced_blend_softlight.frag (hunk 2 above)
Comment (bdero) [reply impeller/entity/shaders/blending/advanced_blend_softlight.frag:8]: Yeah, all of these come from https://www.w3.org/TR/compositing-1/#blendingseparable. Added comments to each of these shaders linking the relevant sections. Documenting this along with the blend situation in general seems like a good idea -- I'll start working on a doc add it in a later patch.

**Pull request merged**